Firestore/core/test/firebase/firestore/remote/datastore_test.mm

```diff
@@ -113,6 +111,52 @@ void Shutdown() {
             "x-google-service: service 2\n");
 }
 
+TEST_F(DatastoreTest, CommitMutationsAuthFailure) {
+  credentials.FailGetToken();
+
+  __block NSError* resulting_error = nil;
+  datastore->CommitMutations(@[], ^(NSError* _Nullable error) {
+    resulting_error = error;
+  });
+  worker_queue.EnqueueBlocking([] {});
+  EXPECT_NE(resulting_error, nil);
+}
+
+TEST_F(DatastoreTest, LookupDocumentsAuthFailure) {
+  credentials.FailGetToken();
+
+  __block NSError* resulting_error = nil;
+  datastore->LookupDocuments(
+      {}, ^(NSArray<FSTMaybeDocument*>* docs, NSError* _Nullable error) {
+        resulting_error = error;
+      });
+  worker_queue.EnqueueBlocking([] {});
+  EXPECT_NE(resulting_error, nil);
+}
+
+/*
+TEST_F(DatastoreTest, AuthWhenDatastoreHasBeenShutDown) {
+  credentials.DelayGetToken();
+  datastore->CommitMutations(@[], ^(NSError* _Nullable error) {
+  FAIL() << "Callback shouldn't be invoked";
+  });
+  Shutdown();
+  EXPECT_NO_THROW(credentials.InvokeGetToken());
+}
+*/
+
+TEST_F(DatastoreTest, AuthOutlivesDatastore) {
+  credentials.DelayGetToken();
+
+  datastore->CommitMutations(@[], ^(NSError* _Nullable error) {
+    FAIL() << "Callback shouldn't be invoked";
+  });
+  Shutdown();
+  datastore.reset();
+
+  EXPECT_NO_THROW(credentials.InvokeGetToken());
```

Comment:
This test currently fails, I'll fix before merging.

Reply:
Hmm, this actually looks like a non-trivial issue.

The root of the problem is that there's an implicit dependency between `grpc::CompletionQueue` and `grpc::ByteBuffer`'s lifetimes. The smallest repro is just:

```cpp
{
  grpc::Slice slice{"foo"};
  grpc::ByteBuffer b{&slice, 1}; // Buffer must be non-empty
  grpc::CompletionQueue cq; // Assuming it's the only gRPC-related object around
} // Once the scope ends, assertion will be triggered, because cq was destroyed before b
```

Details:

In gRPC, C core is initialized once and shut down once. All C++ classes that need the core to be initialized inherit from `GrpcLibraryCodegen`. `GrpcLibraryCodegen` essentially makes C core reference-counted; each constructor increments, and each destructor decrements, the number of references to C core, and once the last reference is destroyed, the C core is shut down.

In this case, destroying `Datastore` destroys `grpc::CompletionQueue`, which happens to be the last reference to C core, so the line 154 (`datastore.reset()`) leads to [global shutdown](https://github.com/grpc/grpc/blob/master/src/core/lib/surface/init.cc#L160-L184). The global shutdown, among other things, [shuts down](https://github.com/grpc/grpc/blob/master/src/core/lib/surface/init.cc#L183) [`ExecCtx`](https://github.com/grpc/grpc/blob/master/src/core/lib/iomgr/exec_ctx.h#L196).

When `EmptyCredentialsProvider::GetToken` is called, the `TokenListener` (a `std::function`) is passed by value, so at the end of the call the destructor of `TokenListener` is called, which leads to the destruction of a lambda created by `Datastore` that contains a [`grpc::ByteBuffer`](https://github.com/firebase/firebase-ios-sdk/blob/master/Firestore/core/src/firebase/firestore/remote/datastore.mm#L153). When a `grpc::ByteBuffer` is destroyed, [it creates an `ExecCtx`](https://github.com/grpc/grpc/blob/master/src/core/lib/surface/byte_buffer.cc#L77), which fails because global shutdown has already been called on `ExecCtx`, leading to an assertion failure and a crash.

(Note that the fact that `GetToken` takes its argument by value isn't really an issue here; if the argument were taken by reference, the problem would surface when the credentials provider is destroyed. The root of the problem seems that the `ByteBuffer`-containing lambda may outlive gRPC core).

Reply:
@wilhuff Re. the above:

1. As far as gRPC is concerned, do you feel it might be a bug? (and hence, worth reporting)
2. I presume we care about this case (Auth outlives Firestore) and don't want a crash there -- let me know if I misunderstand.

Reply:
Submitted an issue to gRPC repo: https://github.com/grpc/grpc/issues/16875